10.0 add delivery carrier seur

jesusVMayor opened the pull request:
https://github.com/OCA/delivery-carrier/pull/127 with a fix and rebased.

Commit:
[ADD] Delivery Carrier Seur Module

angelmoya committed:
MIG delivery_carrier_label_seur

jesusVMayor committed:
[MIG]delivery_carrier_label_seur: Migration to 10.0

rafaelbn commented:
Hi @jesusVMayor , please could you review runbot and travis :red_circle:  ?? thanks a lot!

rafaelbn commented:
Please @jesusVMayor could you review this PR technically and fix bots in order to merge? :smile:

rousseldenis commented on delivery_carrier_label_seur/model/stock.py:
```diff
@@ -0,0 +1,184 @@
+
+    @api.onchange('carrier_id')
+    def carrier_id_change(self):
+        super(StockPicking, self).carrier_id_change()
```
@jesusVMayor You must return result of super as https://github.com/OCA/delivery-carrier/blob/10.0/base_delivery_carrier_label/models/stock_picking.py#L145

rousseldenis commented on delivery_carrier_label_seur/model/stock.py:
```diff
@@ -0,0 +1,184 @@
+    def write(self, vals):
+        res = super(StockPicking, self).write(vals)
+        if 'carrier_id' in vals:
+            self.carrier_id_change()
```
@jesusVMayor I don't think it is a good practice to change just recordset on write(). You are not sure that these changes values are written to db...

Maybe seur_service and product code qhould become compute values ?

rousseldenis commented on delivery_carrier_label_seur/model/delivery.py:
```diff
@@ -0,0 +1,47 @@
+# -*- coding: utf-8 -*-
+# © 2015 FactorLibre - Ismael Calvo <ismael.calvo@factorlibre.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+from openerp import models, fields, api
```
@jesusVMayor 
```suggestion
from odoo import api, fields, models
```

rousseldenis commented on delivery_carrier_label_seur/model/seur_config.py:
```diff
@@ -0,0 +1,20 @@
+# -*- coding: utf-8 -*-
+# © 2015 FactorLibre - Ismael Calvo <ismael.calvo@factorlibre.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+from openerp import models, fields
```
```suggestion
from odoo import fields, models
```

rousseldenis commented on delivery_carrier_label_seur/model/stock.py:
```diff
@@ -0,0 +1,184 @@
+            warehouse.partner_id.country_id.code and \
+            self.partner_id.country_id.code != warehouse.partner_id.\
+                country_id.code:
+            international = True
```
Maybe in a roadmap to a dedicated little module with that field on picking (condition is huge...) ?

rousseldenis commented on .travis.yml:
```diff
@@ -38,6 +38,8 @@ install:
   - pip install unidecode
   - pip install pycountry
   - pip install suds-jurko==0.6
+  - pip install genshi>=0.7
```
Could you add this in manifest external dependencies ?

jesusVMayor replied:
it's a dependency of the seur package, the module don't uses it and if you install the seur package with pip or another toolm genshi it will be installed.

rousseldenis commented on delivery_carrier_label_seur/__manifest__.py:
```diff
@@ -0,0 +1,30 @@
+    "summary": "Allows to generate shipping label for SEUR shipments.",
+    "version": "10.0.1.0.0",
+    "category": "Delivery",
+    "website": "http://factorlibre.com",
```
Could you follow convention in https://github.com/OCA/maintainer-tools/blob/master/template/module/__manifest__.py?

And maybe add some missing fields (e.g.: development_status, maintainers)

rousseldenis commented on delivery_carrier_label_seur/__init__.py:
```diff
@@ -0,0 +1,3 @@
+# -*- coding: utf-8 -*-
+from . import model
```
Could you rename it to models ? https://github.com/OCA/maintainer-tools/tree/master/template/module/models

rousseldenis reviewed (changes requested):
Code review. Some comments

rafaelbn commented:
Hi @jesusVMayor , coudl you review Travis :red_circle: please?

rousseldenis commented on delivery_carrier_label_seur/models/stock.py:
```diff
@@ -0,0 +1,185 @@
+    def write(self, vals):
+        res = super(StockPicking, self).write(vals)
+        if 'carrier_id' in vals:
+            self.carrier_id_change()
```
@jesusVMayor You cannot do that as self can be multi (see Travis error) and carrier_id_change is one

jesusVMayor commented:
Sorry for the delay in the corrections. I have not been able to dedicate much time to this lately. I hope to finish this soon.

rousseldenis commented:
> Sorry for the delay in the corrections. I have not been able to dedicate much time to this lately. I hope to finish this soon.

No problem, just warn when it's ok

jesusVMayor commented:
I added some improvements:

- send labels to printer when is generated as txt (zpl II). It uses report_print_send and queue_job modules.
- Added a button to reprint labels when they are generated as txt.
- Download tracking information with a cron. Disabled by default. Improvement from @zahravelasco
- Removed the external library dependency making the calls to the api directly.
- Check the if the city name correspond to the seur system. If not it shows a wizard to select from the options returned from the api (it could return various options for a unique zip code).

rousseldenis commented:
@jesusVMayor Could you attend last comment and fix travis ?

jesusVMayor commented:
What do you mean? The onchange?  i removed te call from the write method. And the other problems are also resolved.

rousseldenis commented:
> And the other problems are also resolved.

Just the website manifest entry and it's ok

jesusVMayor commented:
Travis green

rousseldenis commented on delivery_carrier_label_seur/models/stock.py:
```diff
@@ -0,0 +1,262 @@
+    def carrier_id_change(self):
+        res = super(StockPicking, self).carrier_id_change()
+        if not self.carrier_id or self.carrier_id.carrier_type != 'seur':
+            return
```
You are not aware what was done on super.
```suggestion
            return res
```

rousseldenis commented on delivery_carrier_label_seur/wizard/zipcode_selector_wizard.py:
```diff
@@ -0,0 +1,39 @@
+        return default_zip_options
+
+    partner_id = fields.Many2one(
+        'res.partner', 'Customer', default=_get_default_partner_id)
```
Maybe

```suggestion
        'res.partner', 'Customer', default=lambda self: self._get_default_partner_id())
```

to allow override ? Same under

rousseldenis reviewed (changes requested):
Code review.

Minor change

rafaelbn commented:
@jesusVMayor are you going to finish this? Do you plan to migrate this to v11? @pedrobaeza what about the OCA CS in Santiago, @liebana or other is going to propose something? @angelmoya ?

jesusVMayor commented on base_delivery_carrier_label/models/stock_picking.py:
```diff
@@ -85,7 +92,7 @@ def get_shipping_label_values(self, label):
             'res_id': self.id,
             'res_model': 'stock.picking',
-            'datas': label['file'].encode('base64'),
+            'datas': b64encode(label['file']),
```
I made a change in the base module that I do not know if it could cause problems for other integrations. But in this case, when we take the zpl label in txt format it is necessary to use b64encode instead of '' .encode ('base64') because this last method fails.
For example:
In a file what ends with:
^FN32^FD1/1^FS
^FN33^FD1.0^FS
^FN34^FD^FS
^FN35^FD^FS
^FN36^FD^FS
^XZ

With ''.encode('base64') the attachment ends with:
^FN32^FD1/1^FS
^FN33^FD1.0^FS
^FN34^FD^FS
^FN35^FD^FS
^FN36^FD^FS
^

I don't know enough of encoding to know the reason. But this change fixed the problem.

pedrobaeza commented:
Closing as we are merging the final revision in https://github.com/OCA/l10n-spain/pull/1382
